demos/object_detection_demo_ssd_async/main.cpp

```diff
@@ -50,523 +135,127 @@ bool ParseAndCheckCommandLine(int argc, char *argv[]) {
-                    prevModeActiveRequestCount--;
-                }
 
-                if (!FLAGS_no_show) {
-                    cv::imshow(imshowWindowTitle, requestResult.frame);
-
-                    const int key = cv::waitKey(1);
-
-                    if (27 == key || 'q' == key || 'Q' == key) {  // Esc
-                        break;
-                    } else if (9 == key) {  // Tab
-                        if (prevModeActiveRequestCount != 0) continue;
-
-                        ExecutionMode prevMode = currentMode;
-                        switchMode(currentMode);
-
-                        {
-                            std::lock_guard<std::mutex> lock(mutex);
-
-                            prevModeActiveRequestCount = prevMode == ExecutionMode::USER_SPECIFIED
-                                                         ? userSpecifiedInferRequests.size() - emptyRequests.size()
-                                                         : 1 - emptyRequests.size();
-                        }
-
-                        putHighlightedText(requestResult.frame, "Switching modes, please wait...",
-                                           cv::Point2f(10, requestResult.frame.rows - 60),
-                                           cv::FONT_HERSHEY_TRIPLEX, 0.75, cv::Scalar(10, 200, 10), 2);
-                        cv::imshow(imshowWindowTitle, requestResult.frame);
-                        cv::waitKey(1);
-
-                        if (prevMode == ExecutionMode::USER_SPECIFIED) {
-                            for (const auto& request : userSpecifiedInferRequests) {
-                                request->Wait(IInferRequest::WaitMode::RESULT_READY);
-                            }
-                        } else minLatencyInferRequest->Wait(IInferRequest::WaitMode::RESULT_READY);
-
-                        {
-                            std::lock_guard<std::mutex> lock(mutex);
-
-                            if (currentMode == ExecutionMode::USER_SPECIFIED) {
-                                emptyRequests.assign(userSpecifiedInferRequests.begin(),
-                                                     userSpecifiedInferRequests.end());
-                            } else emptyRequests = {minLatencyInferRequest};
-                        }
-
-                        if (modeMetrics.find(currentMode) == modeMetrics.end()) {
-                            modeMetrics.insert({currentMode, PerformanceMetrics()});
-                        } else modeMetrics[currentMode] = PerformanceMetrics();
-                    } else {
-                        presenter.handleKey(key);
-                    }
+            //--- Checking for results and rendering data if it's ready
+            //--- If you need just plain data without rendering - check for getProcessedResult() function
+            std::unique_ptr<ResultBase> result;
+            while (result = pipeline.getResult()) {
```

Comment:
Processing results must go before frame reading.

Reply:
why ? The are in cycle anyway. Usually we read-submit-read ready results and process (and submitted frames are processed in parallel thread meanwhile).

Reply:
This will minimize latency for situations when there is a choice to complete one of the frames or start another.

Reply:
they are in cycle. I mean changing order as you described inside the cycle's body of the cycle won't change anything. But if we change the order, it will reduce readability of the code. Please give detailed example how changing the order could help here (in this exact demo) if you insist on this change.

Reply:
When there is a choice to complete one of the frames or start another
You approach postpones postprocessing and reads next frame. This results in increase of latency because of two reasons: 1) already inferred frame waits 2) for some cases you leach CPU resources with postprocessing while IE infers that next frame on CPU.

Another problem is that if postprocessing is slower than reading frames and inference you will end up with `completedInferenceResults` infinitely growing.

Reply:
as we agreed, it's enough to move waitForData between submit and getResult.  Fixed